pkg/server/image_pull.go

```diff
@@ -420,3 +448,68 @@ func getLayers(ctx context.Context, key string, descs []imagespec.Descriptor, va
 	}
 	return
 }
+
+type inProgress struct {
+	entries sync.Map
+}
+
+func newInProgress() *inProgress {
+	return &inProgress{}
+}
+
+func (inp *inProgress) load(key string) (*imageBroadCaster, bool) {
+	val, ok := inp.entries.Load(key)
+	if ok {
+		return val.(*imageBroadCaster), true
+	}
+	return nil, false
+}
+
+func (inp *inProgress) store(key string) *imageBroadCaster {
+	bc := newImageBroadcaster()
+	inp.entries.Store(key, bc)
+	return bc
+}
+
+func (inp *inProgress) delete(key string) {
+	inp.entries.Delete(key)
+}
+
+type imageBroadCaster struct {
+	c     *sync.Cond
+	image containerd.Image
+	done  bool
+	err   error
+}
+
+func newImageBroadcaster() *imageBroadCaster {
+	return &imageBroadCaster{
+		c: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (inp *inProgress) done(key string, err error, image containerd.Image) {
+	broadcaster, ok := inp.load(key)
+	if ok {
+		broadcaster.c.L.Lock()
+		broadcaster.done = true
+		broadcaster.image = image
+		broadcaster.err = err
+		broadcaster.c.Broadcast()
```

Comment:
Curious on thoughts about a slice of channels and fanout to them, or just stick with the sync.Cond/broadcast approach

Reply:
Not sure I follow but can this synchronization be accomplished with just a mutex? The winner locks during the pull and the losers lock to check the result.

Reply:
We already have synchronization at the map read/write layer just by virtue of using `sync.Map`. My question was mainly this whole `sync.Cond` stuff is just a nice builtin stdlib way to alert a bunch of waiters at one time and let them all read one piece of data (the `imageBroadcaster` in the map) after they've been woken up. I was wondering if anyone thought there may be a cleaner approach using go channels here as they block by design until a value is sent down them. I don't think the current approach is too ugly or unwieldy, just if there's a more idiomatic way for this I'd prefer that. This problem seems to fit `sync.Cond` nicely though.

Reply:
I agree Cond Broadcast is the way to wake multiple waiters and is fine here. Just saying it could be accomplished by just a mutex with the difference being that with a mutex only one waiter runs at a time.